Description: Add configure option to reduce visual clutter at boot time
 If this option is enabled, then do all of the following:
 .
 Don't display introductory message about line editing unless we're actually
 offering a shell prompt.  (This is believed to be a workaround for a
 different bug.  We'll go with this for now, but will drop this in favour of
 a better fix upstream if somebody figures out what that is.)
 .
 Don't clear the screen just before booting if we never drew the menu in the
 first place.
 .
 Remove verbose messages printed before reading configuration.  In some ways
 this is awkward because it makes debugging harder, but it's a requirement
 for a smooth-looking boot process; we may be able to do better in future.
 Upstream doesn't want this, though.
 .
 Disable the cursor as well, for similar reasons of tidiness.
 .
 Suppress kernel/initrd progress messages, except in recovery mode.
 .
 Suppress "GRUB loading" message unless Shift is held down.  Upstream
 doesn't want this, as it makes debugging harder.  Ubuntu wants it to
 provide a cleaner boot experience.
Author: Colin Watson <cjwatson@ubuntu.com>
Bug-Ubuntu: https://bugs.launchpad.net/bugs/386922
Bug-Ubuntu: https://bugs.launchpad.net/bugs/861048
Forwarded: (partial) http://lists.gnu.org/archive/html/grub-devel/2009-09/msg00056.html
Last-Update: 2013-08-15

diff -Naur grub-2.00/configure.ac grub-2.00.tpg/configure.ac
--- grub-2.00/configure.ac	2012-06-26 01:47:37.000000000 +0000
+++ grub-2.00.tpg/configure.ac	2013-08-23 16:01:17.717262663 +0000
@@ -1076,6 +1076,16 @@
 AC_SUBST([LIBZFS])
 AC_SUBST([LIBNVPAIR])
 
+AC_ARG_ENABLE([quiet-boot],
+              [AS_HELP_STRING([--enable-quiet-boot],
+                              [emit fewer messages at boot time (default=no)])],
+              [], [enable_quiet_boot=no])
+if test x"$enable_quiet_boot" = xyes ; then
+  AC_DEFINE([GRUB_QUIET_BOOT], [1],
+            [Define to 1 to make GRUB quieter at boot time.])
+fi
+AC_SUBST([enable_quiet_boot])
+
 LIBS=""
 
 AC_SUBST([FONT_SOURCE])
@@ -1232,5 +1242,10 @@
 else
 echo With libzfs support: No "($libzfs_excuse)"
 fi
+if [ x"$enable_quiet_boot" = xyes ]; then
+echo With quiet boot: Yes
+else
+echo With quiet boot: No
+fi
 echo "*******************************************************"
 ]
diff -Naur grub-2.00/grub-core/boot/i386/pc/boot.S grub-2.00.tpg/grub-core/boot/i386/pc/boot.S
--- grub-2.00/grub-core/boot/i386/pc/boot.S	2011-10-19 18:46:09.000000000 +0000
+++ grub-2.00.tpg/grub-core/boot/i386/pc/boot.S	2013-08-23 16:03:18.874615544 +0000
@@ -19,6 +19,9 @@
 
 #include <grub/symbol.h>
 #include <grub/machine/boot.h>
+#ifdef GRUB_QUIET_BOOT
+#include <grub/machine/memory.h>
+#endif
 
 /*
  *  defines for the code go here
@@ -150,9 +153,17 @@
 	/* save drive reference first thing! */
 	pushw	%dx
 
+#ifdef GRUB_QUIET_BOOT
+	/* is either shift key held down? */
+	movw	$(GRUB_MEMORY_MACHINE_BIOS_DATA_AREA_ADDR + 0x17), %bx
+	andb	$3, (%bx)
+	jz	2f
+#endif
+
 	/* print a notification message on the screen */
 	MSG(notification_string)
 
+2:
 	/* set %si to the disk address packet */
 	movw	$disk_address_packet, %si
 
diff -Naur grub-2.00/grub-core/boot/i386/pc/diskboot.S grub-2.00.tpg/grub-core/boot/i386/pc/diskboot.S
--- grub-2.00/grub-core/boot/i386/pc/diskboot.S	2012-06-23 09:10:37.000000000 +0000
+++ grub-2.00.tpg/grub-core/boot/i386/pc/diskboot.S	2013-08-23 16:03:30.145926383 +0000
@@ -18,6 +18,9 @@
 
 #include <grub/symbol.h>
 #include <grub/machine/boot.h>
+#ifdef GRUB_QUIET_BOOT
+#include <grub/machine/memory.h>
+#endif
 
 /*
  *  defines for the code go here
@@ -25,6 +28,12 @@
 
 #define MSG(x)	movw $x, %si; call LOCAL(message)
 
+#ifdef GRUB_QUIET_BOOT
+#define SILENT(x)	call LOCAL(check_silent); jz LOCAL(x)
+#else
+#define SILENT(x)
+#endif
+
 	.file	"diskboot.S"
 
 	.text
@@ -50,11 +59,14 @@
 	/* save drive reference first thing! */
 	pushw	%dx
 
+	SILENT(after_notification_string)
+
 	/* print a notification message on the screen */
 	pushw	%si
 	MSG(notification_string)
 	popw	%si
 
+LOCAL(after_notification_string):
 	/* this sets up for the first run through "bootloop" */
 	movw	$LOCAL(firstlist), %di
 
@@ -279,7 +291,10 @@
 	/* restore addressing regs and print a dot with correct DS
 	   (MSG modifies SI, which is saved, and unused AX and BX) */
 	popw	%ds
+	SILENT(after_notification_step)
 	MSG(notification_step)
+
+LOCAL(after_notification_step):
 	popa
 
 	/* check if finished with this dataset */
@@ -295,8 +310,11 @@
 /* END OF MAIN LOOP */
 
 LOCAL(bootit):
+	SILENT(after_notification_done)
 	/* print a newline */
 	MSG(notification_done)
+
+LOCAL(after_notification_done):
 	popw	%dx	/* this makes sure %dl is our "boot" drive */
 	ljmp	$0, $(GRUB_BOOT_MACHINE_KERNEL_ADDR + 0x200)
 
@@ -320,6 +338,14 @@
 /* go here when you need to stop the machine hard after an error condition */
 LOCAL(stop):	jmp	LOCAL(stop)
 
+#ifdef GRUB_QUIET_BOOT
+LOCAL(check_silent):
+	/* is either shift key held down? */
+	movw	$(GRUB_MEMORY_MACHINE_BIOS_DATA_AREA_ADDR + 0x17), %bx
+	andb	$3, (%bx)
+	ret
+#endif
+
 notification_string:	.asciz "loading"
 
 notification_step:	.asciz "."
diff -Naur grub-2.00/grub-core/kern/main.c grub-2.00.tpg/grub-core/kern/main.c
--- grub-2.00/grub-core/kern/main.c	2012-06-25 20:33:41.000000000 +0000
+++ grub-2.00.tpg/grub-core/kern/main.c	2013-08-23 16:04:01.200783273 +0000
@@ -207,13 +207,23 @@
 void __attribute__ ((noreturn))
 grub_main (void)
 {
+#ifdef GRUB_QUIET_BOOT
+  struct grub_term_output *term;
+#endif
+
   /* First of all, initialize the machine.  */
   grub_machine_init ();
 
+#ifdef GRUB_QUIET_BOOT
+  /* Disable the cursor until we need it.  */
+  FOR_ACTIVE_TERM_OUTPUTS(term)
+    grub_term_setcursor (term, 0);
+#else
   /* Hello.  */
   grub_setcolorstate (GRUB_TERM_COLOR_HIGHLIGHT);
   grub_printf ("Welcome to GRUB!\n\n");
   grub_setcolorstate (GRUB_TERM_COLOR_STANDARD);
+#endif
 
   /* Load pre-loaded modules and free the space.  */
   grub_register_exported_symbols ();
@@ -232,5 +242,12 @@
 
   grub_load_config ();
   grub_load_normal_mode ();
+
+#ifdef GRUB_QUIET_BOOT
+  /* If we have to enter rescue mode, enable the cursor again.  */
+  FOR_ACTIVE_TERM_OUTPUTS(term)
+    grub_term_setcursor (term, 1);
+#endif
+
   grub_rescue_run ();
 }
diff -Naur grub-2.00/grub-core/kern/rescue_reader.c grub-2.00.tpg/grub-core/kern/rescue_reader.c
--- grub-2.00/grub-core/kern/rescue_reader.c	2012-01-31 21:53:43.000000000 +0000
+++ grub-2.00.tpg/grub-core/kern/rescue_reader.c	2013-08-23 16:04:06.511930639 +0000
@@ -77,7 +77,9 @@
 void __attribute__ ((noreturn))
 grub_rescue_run (void)
 {
+#ifdef GRUB_QUIET_BOOT
   grub_printf ("Entering rescue mode...\n");
+#endif
 
   while (1)
     {
diff -Naur grub-2.00/grub-core/normal/main.c grub-2.00.tpg/grub-core/normal/main.c
--- grub-2.00/grub-core/normal/main.c	2012-06-26 01:59:57.000000000 +0000
+++ grub-2.00.tpg/grub-core/normal/main.c	2013-08-23 16:04:28.568537427 +0000
@@ -399,6 +399,15 @@
 grub_normal_read_line_real (char **line, int cont, int nested)
 {
   const char *prompt;
+#ifdef GRUB_QUIET_BOOT
+  static int displayed_intro;
+
+  if (! displayed_intro)
+    {
+      grub_normal_reader_init (nested);
+      displayed_intro = 1;
+    }
+#endif
 
   if (cont)
     /* TRANSLATORS: it's command line prompt.  */
@@ -446,7 +455,9 @@
       return;
     }
 
+#ifndef GRUB_QUIET_BOOT
   grub_normal_reader_init (nested);
+#endif
 
   while (1)
     {
diff -Naur grub-2.00/grub-core/normal/menu.c grub-2.00.tpg/grub-core/normal/menu.c
--- grub-2.00/grub-core/normal/menu.c	2012-05-17 12:55:37.000000000 +0000
+++ grub-2.00.tpg/grub-core/normal/menu.c	2013-08-23 16:04:35.616731837 +0000
@@ -669,12 +669,18 @@
 
 /* Callback invoked immediately before a menu entry is executed.  */
 static void
-notify_booting (grub_menu_entry_t entry,
+notify_booting (grub_menu_entry_t entry
+#ifdef GRUB_QUIET_BOOT
+		__attribute__((unused))
+#endif
+		,
 		void *userdata __attribute__((unused)))
 {
+#ifndef GRUB_QUIET_BOOT
   grub_printf ("  ");
   grub_printf_ (N_("Booting `%s'"), entry->title);
   grub_printf ("\n\n");
+#endif
 }
 
 /* Callback invoked when a default menu entry executed because of a timeout
@@ -722,6 +728,9 @@
       int boot_entry;
       grub_menu_entry_t e;
       int auto_boot;
+#ifdef GRUB_QUIET_BOOT
+      int initial_timeout = grub_menu_get_timeout ();
+#endif
 
       boot_entry = run_menu (menu, nested, &auto_boot);
       if (boot_entry < 0)
@@ -731,7 +740,11 @@
       if (! e)
 	continue; /* Menu is empty.  */
 
-      grub_cls ();
+#ifdef GRUB_QUIET_BOOT
+      /* Only clear the screen if we drew the menu in the first place.  */
+      if (initial_timeout != 0)
+#endif
+	grub_cls ();
 
       if (auto_boot)
 	grub_menu_execute_with_fallback (menu, e, autobooted,
diff -Naur grub-2.00/util/grub.d/10_linux.in grub-2.00.tpg/util/grub.d/10_linux.in
--- grub-2.00/util/grub.d/10_linux.in	2013-08-23 15:59:44.000000000 +0000
+++ grub-2.00.tpg/util/grub.d/10_linux.in	2013-08-23 15:58:25.000000000 +0000
@@ -20,6 +20,7 @@
 prefix="@prefix@"
 exec_prefix="@exec_prefix@"
 datarootdir="@datarootdir@"
+enable_quiet_boot="@enable_quiet_boot@"
 
 . "@datadir@/@PACKAGE@/grub-mkconfig_lib"
 
@@ -132,18 +133,22 @@
     fi
     printf '%s\n' "${prepare_boot_cache}" | sed "s/^/$submenu_indentation/"
   fi
+  if [ x"$enable_quiet_boot" = xno ] || [ x"$type" != xsimple ]; then
   message="$(gettext_printf "Loading Linux %s ..." ${version})"
   sed "s/^/$submenu_indentation/" << EOF
 	echo	'$message'
 	linux  ${rel_dirname}/${basename} root=${linux_root_device_thisversion} ro ${args} resume=`swapon -s | grep '/dev/sd.[0-9]\{1,\}' -o | if read line; then echo "$line"; fi` init=/lib/systemd/systemd
 EOF
+  fi
   if test -n "${initrd}" ; then
     # TRANSLATORS: ramdisk isn't identifier. Should be translated.
+    if [ x"$enable_quiet_boot" = xno ] || [ x"$type" != xsimple ]; then
     message="$(gettext_printf "Loading initial ramdisk ...")"
     sed "s/^/$submenu_indentation/" << EOF
 	echo	'$message'
 	initrd	${rel_dirname}/${initrd}
 EOF
+    fi
   fi
   sed "s/^/$submenu_indentation/" << EOF
 }
